### PR TITLE
Mod 1e9+7 Tutorial

### PR DESCRIPTION
**Pull request by tannudaral**
Please let me know if any changes are required.

### CODE REVIEW

File: tutorials/basic-topics/mod.md [hunk 1]
```diff
@@ -1,10 +1,83 @@
...
   - MOD
----
\ No newline at end of file
+---
+
+<TutorialAuthors names="@tannudaral"/>
+
+## Overview
+
+Sometimes, the answer to a problem is a very large number but it is enough to output it "modulo m", i.e., the remainder when the answer is divided by m (for example, "modulo 1e9 + 7"). The idea is that even if the actual answer is very large, it suffices to use the data types int and long long. Many languages do not support large-integer arithmetic, so this method avoids integer overflow.
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:16]: I think this intro is from CP handbook. If so, don't copy the exact content from other sources, use your own words to describe it.

File: tutorials/basic-topics/mod.md [hunk 2]
```diff
@@ -1,10 +1,83 @@
...
+
+## MOD
+
+The task of modulo operator `%`, also know as the remainder operator, is to give the remainder. We denote by x mod m, the remainder when x is divided by m. For example, $17\,mod\,5= 2$ because 17 = 3*5 + 2.
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:20]: Don't use backtick. Use LaTex. Apply to all.

File: tutorials/basic-topics/mod.md [hunk 3]
```diff
@@ -1,10 +1,83 @@
...
+
+### Modular Arithmetic   
+An important property of the modulo is that in addition, subtraction and multiplication, the remainder can be taken before the operation:  
+#### 1. Addition
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:24]: add a line break before L24

File: tutorials/basic-topics/mod.md [hunk 4]
```diff
@@ -1,10 +1,83 @@
...
+
+### Modular Arithmetic   
+An important property of the modulo is that in addition, subtraction and multiplication, the remainder can be taken before the operation:  
+#### 1. Addition  
+$(a + b)\,mod\,m$ $=$ $(a\,mod\,m + b\,mod\,m)\,mod\,m$
+
+#### 2. Subtraction  
+Usually we want the remainder to always be between 0...m−1. However, in C++ and other languages, the remainder of a negative number is either zero or negative. An easy way to make sure there are no negative remainders is to add m to the result. This is only needed when there are subtractions in the code and the remainder may become negative. 
+
+  $(a\,−\,b)\,mod\,m$ $=$ $(a \, mod\, m \,−\, b \,mod \,m\, + \,m)\, mod \,m$
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:30]: remove the extra space at the beginning

File: tutorials/basic-topics/mod.md [hunk 5]
```diff
@@ -1,10 +1,83 @@
...
+### Example 
+Let's understand this with the factorial of a number program. The following code calculates n!, the factorial of n, modulo m:
+
+```cpp
+int factorial(int n){
+    int M = 1e9+7;
+    long long fact = 1;
+    for (int i = 2; i <= n; ++i)
+        fact = fact * i;    // WRONG APPROACH
+                            // Here, fact may exceed 2^64-1
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:60]: put the comment above that line

File: tutorials/basic-topics/mod.md [hunk 6]
```diff
@@ -1,10 +1,83 @@
...
+```
+Thus, we can take the remainder after every operation and the numbers will never become too large.
+
+```cpp
+int factorial(int n){
+    int M = 1e9+7;
+    long long fact = 1;
+    for (int i = 2; i <= n; ++i)
+        fact = (fact * i) % M;    // Here, fact never
+                                  // exceeds 10^9+7
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:73]: put the comment above that line

File: tutorials/basic-topics/mod.md [hunk 7]
```diff
@@ -1,10 +1,83 @@
...
+Hence, we can calculate the modular multiplicative inverse $a^{−1}$ using $a^{−1} = a^{m−2}\, mod\, m$ when $m$ is prime. We can now define the division operator as:
+
+$(a\, /\, b) \,mod\, m$ $=$ $(a\, mod\, m\, * b^{-1} \,mod \,m)\, mod\, m$  
+
+### Example 
+Let's understand this with the factorial of a number program. The following code calculates n!, the factorial of n, modulo m:
+
+```cpp
+int factorial(int n){
+    int M = 1e9+7;
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:56]: add space. `1e9 + 7`

File: tutorials/basic-topics/mod.md [hunk 8]
```diff
@@ -1,10 +1,83 @@
...
+                            // Here, fact may exceed 2^64-1
+ 
+    return fact%M;
+}
+```
+Thus, we can take the remainder after every operation and the numbers will never become too large.
+
+```cpp
+int factorial(int n){
+    int M = 1e9+7;
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:69]: add space. `1e9 + 7`

File: tutorials/basic-topics/mod.md [hunk 9]
```diff
@@ -1,10 +1,83 @@
...
+    long long fact = 1;
+    for (int i = 2; i <= n; ++i)
+        fact = fact * i;    // WRONG APPROACH
+                            // Here, fact may exceed 2^64-1
+ 
+    return fact%M;
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:62]: add space. i.e. `fact % M`

File: tutorials/basic-topics/mod.md [hunk 10]
```diff
@@ -1,10 +1,83 @@
...
+    for (int i = 2; i <= n; ++i)
+        fact = (fact * i) % M;    // Here, fact never
+                                  // exceeds 10^9+7
+ 
+    return fact;
+}
+```
+
+## Why 1e9+7?
+
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:80]: space for 1e9 + 7

File: tutorials/basic-topics/mod.md [hunk 11]
```diff
@@ -1,10 +1,83 @@
...
+Fermat's little theorem says that as long as the modulus m is a prime number ($10^9+7$ is prime) then $a^{m}\, mod\, m = a\, mod\, m$. Working backwards, $a^{m−1}\, mod\, m = a^{0}\, mod\, m = 1 = a ⋅ a^{m−2}\, mod\, m$, therefore the number we need is $a^{m−2}\, mod\, m$.
+
+Hence, we can calculate the modular multiplicative inverse $a^{−1}$ using $a^{−1} = a^{m−2}\, mod\, m$ when $m$ is prime. We can now define the division operator as:
+
+$(a\, /\, b) \,mod\, m$ $=$ $(a\, mod\, m\, * b^{-1} \,mod \,m)\, mod\, m$  
+
+### Example 
+Let's understand this with the factorial of a number program. The following code calculates n!, the factorial of n, modulo m:
+
+```cpp
```
Comment (wingkwong) [inline tutorials/basic-topics/mod.md:54]: Use `<Tabs/>` for code.

See [here](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#solutions)

**Review by wingkwong — CHANGES_REQUESTED** [review]
If you've referenced some external sites, state them at the end.

File: tutorials/basic-topics/mod.md (hunk 10 above)
Comment (vigneshshiv) [reply tutorials/basic-topics/mod.md:93]: Space not added yet.

File: tutorials/basic-topics/mod.md [hunk 12]
```diff
@@ -1,10 +1,101 @@
...
+
+<TutorialAuthors names="@tannudaral"/>
+
+## Overview
+
+When the answer to a problem is a very large number, problem setters expect you to output it "modulo m", that is, the remainder after dividing the answer by $m$ (for example, "modulo 1e9 + 7"). So if the actual answer is very large, with the use of modulo $m$, it would be sufficient to use the data types int and long long. Since many languages do not support large-integer arithmetic, this method avoids integer overflow.
+
+## MOD
+
+The task of modulo operator $\%$, also know as the remainder operator, is to give the remainder. We denote it by $x\,mod\,m$, the remainder when $x$ is divided by $m$. For example, 17 mod 5 = 2 because 17 = 3*5 + 2.
```
Comment (vigneshshiv) [inline tutorials/basic-topics/mod.md:20]: Use LaTeX when you specify numbers

**Review by vigneshshiv — COMMENTED** [review]
Overall looks good, Few changes needed, go through the comments

**Comment by tannudaral** [comment]
Made all requested changes. Pls check!

File: tutorials/basic-topics/mod.md [hunk 13]
```diff
@@ -1,10 +1,101 @@
...
+$(a\, /\, b) \,mod\, m$ $=$ $(a\, mod\, m\, * b^{-1} \,mod \,m)\, mod\, m$  
+
+### Example 
+Let's understand this with the factorial of a number program. The following code calculates $n!$, the factorial of $n$, modulo $m$:
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+int factorial(int n){
```
Comment (vigneshshiv) [inline tutorials/basic-topics/mod.md:58]: Leave space between **) {**

File: tutorials/basic-topics/mod.md [hunk 14]
```diff
@@ -1,10 +1,101 @@
...
+</TabItem>
+</Tabs>
+
+Thus, we can take the remainder after every operation and the numbers will never become too large.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+int factorial(int n){
```
Comment (vigneshshiv) [inline tutorials/basic-topics/mod.md:78]: Leave space between **) {**

**Review by vigneshshiv — COMMENTED** [review]
All good, one more round final changes.

File: tutorials/basic-topics/mod.md (hunk 14 above)
Comment (tannudaral) [reply tutorials/basic-topics/mod.md]: Done. Pls check!

**Review by vigneshshiv — APPROVED** [review]
Looks good.

**Review by wingkwong — APPROVED** [review]
Overall it looks ok. Will fine tune a bit myself.

**Comment by wingkwong** [comment]
@tannudaral cleaned up a bit and published. Thank you very much!

**Pull request merged**